### ISSUE TITLE
Sensor start_date gets updated for the same try in reschedule mode

### ISSUE THREAD

**Issue by wolfier** [apache/airflow#15645]
**Apache Airflow version**: 2.0.0

**What happened**:

Sensor start_date gets updated to each time the task instance is rescheduled for the same try.

**What you expected to happen**:

I expect the `start_date` of a sensor task instance to remain the same within a given try. The `start_date` should be the start of first poke instead of the start time of the following rescheduled attempts. 

This behaviour does not match the description in the [source code](https://github.com/apache/airflow/blob/2.0.0/airflow/models/taskinstance.py#L982-L983).

>             # Set the task start date. In case it was re-scheduled use the initial
>             # start date that is recorded in task_reschedule table

Airflow did behave as expected in 1.10.x but it stopped in 2.0+.

**How to reproduce it**:

Run any sensors and watch the start_date value change at after each retry.

```python
with DAG(
        dag_id="check_dag",
        start_date=datetime(2021, 1, 1),
        schedule_interval='@daily',
        catchup=False,
) as dag:
    p = ExternalTaskSensor(
        task_id='sensor_task',
        external_dag_id='my_dag',
        mode='reschedule'
    )
```

**Comment by eladkal** [comment]
hi @wolfier are you still working on this issue?

**Comment by o-nikolas** [comment]
Given the discussion in #20224 it seems the most doable solution is to undo the optimization in #9087 what do folks think about that proposal? Any other reasonable ideas?

Also I tested this on main, and the issue is indeed still present (as one would expect), so updating the label (re [this triage initiative](https://apache-airflow.slack.com/archives/CCPRP7943/p1663586469433659))

**Comment by potiuk** [comment]
Yes. I think it should be - for consistentcy - fixed.  Not sure if "undoing" the optimisation is a good approach, more like fixing it so that start_date is not updated in this case.

**Comment by amoghrajesh** [comment]
@potiuk @eladkal I see that this issue did not get completed. If given some context/reference, I would like to try it

**Comment by potiuk** [comment]
I think all the context is in the issue description/comments :)
